lightningd/options.c

```diff
@@ -449,6 +449,10 @@ static void setup_default_config(struct lightningd *ld)
 		ld->config = testnet_config;
 	else
 		ld->config = mainnet_config;
+
+        /* Set default PID file name to be per-network */
+        tal_free(ld->pidfile);
+        ld->pidfile = tal_fmt(ld, "lightningd-%s.pid", get_chainparams(ld)->network_name);
```

Comment:
This doesn't work, though, if they override it later with `--network=`.  This is why we need a new function which figures it on-demand say:

```
const char *get_pidfile(const tal_t *ctx, struct lightningd *ld)
{
        if (ld->pidfile)
                return ld->pidfile;
        return tal_fmt(ctx, "lightningd-%s.pid", get_chainparams(ld)->network_name);
}
```

Then call it from `pidfile_create()` instead of accessing ld->pidfile directly, and also from a new show function:

```
static char *opt_show_pidfile(char buf[OPT_SHOW_LEN], struct lightningd *ld)
{
        char *pidfile = get_pidfile(NULL, ld);
        snprintf(buf, OPT_SHOW_LEN, pidfile);
        tal_free(pidfile);
        return NULL;
}
```

And yeah, that means you need your code code back that I nacked, to hand the `add_config()` case, sorry!

Reply:
Hi,

In testing it seems to work. `setup_default_config(struct lightningd *ld)` is run twice in `lightningd/options.c` The first time it is run `ld->pidfile` is set to "lightningd-testnet.pid" (because testnet is the default network). The second time it is run after options have been parsed so network is set:
```c
        opt_early_parse(argc, argv, config_log_stderr_exit);
        /* Now we can set up defaults, depending on whether testnet or not */
        setup_default_config(ld);
```
I have tested using the `--network=bitcoin` option and the PID file `lightningd-bitcoin.pid` is indeed created.

In comments above the fact that `setup_default_config` runs twice was causing a memory leak in `setup_default_config` which was fixed by adding:
```c
tal_free(ld->pidfile);
```

We'll get there eventually!!! :-)

Reply:
static char *opt_show_pidfile(char buf[OPT_SHOW_LEN], struct lightningd *ld)
    {
            char *pidfile = get_pidfile(NULL, ld);
            snprintf(buf, OPT_SHOW_LEN, pidfile);
            tal_free(pidfile);
            return NULL;
    }

For this code, if `ld->pidfile != NULL`, would it not erroenously free `ld->pidfile`?  Or is `ld->pidfile == NULL` assuredly in this case? (in which case we should `assert(ld->pidfile == NULL);`)

Reply:
Good point, that code was buggy :)